### PR TITLE
nixos: fix systemd-boot security warnings

### LINKED ISSUE #279362: NixOS with systemd-boot results in "security hole" warnings
### Describe the bug
Since NixOS 23.11, using `boot.loader.systemd-boot.enable = true;` results in these warnings at bootloader install time:

```
⚠️ Mount point '/boot' which backs the random seed file is world accessible, which is a security hole! ⚠️
⚠️ Random seed file '/boot/loader/.#bootctlrandom-seedc878009c19a876dc' is world accessible, which is a security hole! ⚠️
```

(Minimal repro: `sudo bootctl --esp-path=/mnt/boot install`)

This happens when following the official NixOS installation manual (https://nixos.org/manual/nixos/stable/#sec-installation-manual-partitioning), which in particular says that for UEFI systems the ESP is to be created with `mkfs.fat -F 32 -n boot /dev/sda3` and mounted with `mount /dev/disk/by-label/boot /mnt/boot`. Doing that makes files under `/mnt/boot` world readable, which apparently is a security hole.

I fixed the install time warning by mounting the ESP with `sudo mount -o umask=007 /dev/sda3 /mnt/boot`, but since `nixos-generate-config` doesn't pick up on mount options (at least not the `umask=` one), the installed OS still has the security hole and the warnings will be shown the next time the bootloader gets updated.

Manually adding the required `umask=` mount option in hardware-configuration.nix should permanently fix it.

### Steps To Reproduce
1. Boot NixOS live medium in UEFI mode.
2. Follow installation manual: https://nixos.org/manual/nixos/stable/#sec-installation-manual-partitioning.
3. Set `boot.loader.systemd-boot.enable = true;` in configuration.nix.
4. Run `nixos-install ...` and see the mentioned warning.

### Expected behavior
* Installation instructions don't result in a system with security holes.
* `nixos-generate-config` should probably pick up on umask= mount options, so that when the installation instructions get fixed the fix is also propagated to the installed OS.

### Metadata
NixOS 23.11 (32f63574c85fbc80e4ba1fbb932cde9619bad25e).

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by bjornfor**
## Description of changes

- nixos-generate-config: preserve vfat filesystem mount permissions
- nixos/doc: suggest mounting the ESP on /boot with umask=077

Fixes https://github.com/NixOS/nixpkgs/issues/279362.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/installer/tools/nixos-generate-config.pl [hunk 1]
```diff
@@ -453,6 +453,17 @@ sub in {
         }
     }
 
+    # Preserve umask (fmask, dmask) settings for vfat filesystems.
+    # (The default is to mount these world-readable, but that's a security risk
+    # for the EFI System Partition.)
+    if ($fsType eq "vfat") {
+        for (@superOptions) {
+            if ($_ =~ /fmask|dmask/) {
```
Comment (ncfavier) [inline nixos/modules/installer/tools/nixos-generate-config.pl:461]:
```suggestion
            if ($_ =~ /fmask|dmask|umask/) {
```

Comment (ncfavier) [reply nixos/modules/installer/tools/nixos-generate-config.pl:461]: Oh I guess that can't occur there, never mind.

Comment (bjornfor) [reply nixos/modules/installer/tools/nixos-generate-config.pl:461]: Yes, it seems so. (I started looking for `umask`, then noticed that `/proc/self/mountinfo` represents `umask` as  `fmask` + `dmask`.)

File: nixos/modules/installer/tools/nixos-generate-config.pl [hunk 2]
```diff
@@ -453,6 +453,17 @@ sub in {
         }
     }
 
+    # Preserve umask (fmask, dmask) settings for vfat filesystems.
+    # (The default is to mount these world-readable, but that's a security risk
+    # for the EFI System Partition.)
+    if ($fsType eq "vfat") {
```
Comment (Aleksanaa) [inline nixos/modules/installer/tools/nixos-generate-config.pl:459]: This condition is somewhat strange. Maybe we need to make it more precise or more general?

**Review by Aleksanaa — COMMENTED** [review]
Diff lgtm. A release note may be needed.

File: nixos/modules/installer/tools/nixos-generate-config.pl (hunk 2 above)
Comment (bjornfor) [reply nixos/modules/installer/tools/nixos-generate-config.pl:459]: What do you have in mind? And, can we do that in a follow-up PR?

**Comment by bjornfor** [comment]
> Diff lgtm. A release note may be needed.

I consider this a bugfix for a regression since release-23.11, and I want to backport it to release-23.11. Not sure how a release note fits in there.

File: nixos/modules/installer/tools/nixos-generate-config.pl (hunk 2 above)
Comment (bjornfor) [reply nixos/modules/installer/tools/nixos-generate-config.pl:459]: In what way is the condition strange?

Comment (Aleksanaa) [reply nixos/modules/installer/tools/nixos-generate-config.pl:459]: vfat doesn't necessarily mean EFI partition. Also do partitions in other formats need this feature?

Comment (bjornfor) [reply nixos/modules/installer/tools/nixos-generate-config.pl:459]: Ah, I see. I thought about checking the mountpoint, but then again I didn't want to introduce weird/artificial restrictions in this tool either.

Having nixos-generate-config emit fmask= and dmask= on non-ESP mounts isn't necessarily a bad thing either. One could argue it's the correct thing to do in any case (preserving the permissions that was used at install time). The only thing I could think of was if we could detect when no umask was used for the mount and then _not_ emit fmask= and dmask= in hardware-configuration.nix. But I'm not sure where the default mount option should be read from, and so I don't think it's worth doing.

> Also do partitions in other formats need this feature?

I don't know. But this was specifically a fix for systemd-boot/EFI, and it only supports FAT32. (Also, we can always improve later.)

**Review by ElvishJerricco — APPROVED** [review]
I like this fix. I think it could be argued that we should be preserving *any* mount options that the user set when they mounted *any* file system, but that's a much more difficult feature (since a lot of the mount options you get from the mount table are automatic ones). So I think using this as a starting point is a good idea.

**Pull request merged**